### PR TITLE
Refactor: Remove err() argument and support provision for error handling

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**

### CODE REVIEW

**Comment by certik** [comment]
I think this is problematic. See https://github.com/lfortran/lfortran/pull/3262.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -4731,16 +4731,12 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                         is_function = true;
                         return resolve_intrinsic_function(x.base.base.loc, var_name);
                     } else {
-                        tmp = create_func(al, x.base.base.loc, args,
-                            [&](const std::string &msg, const Location &loc) {
-                                throw SemanticError(msg, loc); });
+                        tmp = create_func(al, x.base.base.loc, args);
```
[the remark this answers is not in the document]
Comment (certik) [reply src/lfortran/semantics/ast_common_visitor.h:4734]: Here is a way to do it:
```c++
tmp = create_func(al, x.base.base.loc, args, diagnostics);
if (tmp == nullptr) throw SemanticAbort(); // The error will already be in `diagnostics`
```

[the remark this answers is not in the document]
Comment (Shaikh-Ubaid) [reply src/lfortran/semantics/ast_common_visitor.h:4734]: Sure, I will update this after merging/completion of https://github.com/lfortran/lfortran/pull/3313.

**Review by certik — APPROVED** [review]
I think that this is good, thanks for cleaning this up.

**Pull request merged**